### DISCUSSION TITLE
feature/player-ui Branch Beta Release?

### DISCUSSION THREAD

**Discussion by Insan1ty0ne** [tamius-han/ultrawidify#190] [General]
I have been using your plugin for YEARS and absolutely love it. I thought development had stopped on the main version as there haven't been any new releases since 2021. However, I saw that a lot of work has been going on over at the feature/player-ui branch. Will this be a big update to Ultrawidify? Depending on where this branch is at in the development process, perhaps building and uploading a beta release for us to test / play around with might be something worth thinking about? Especially since your website for nightly builds has been down for such a long time.

**Comment by tamius-han** [comment]
Yeah, that's going to be a bigger UI update. However, it's progressing _very_ slowly (+ there's been at least two hard resets of the UI redesign). It turns out that injecting code into a webpage is a bit of a pain in the ass when you're doing that from an extension. 

The current stuff from UI branch is not very beta-ready. It's early to mid alpha at best: the new UI is currently set to always be visible. There are plans to have new UI only trigger when:
* hovering over a certain region of the screen. Trigger area will be visible while moving mouse over video, but will vanish if mouse is kept still for some period of time (1-3s) 
* hotkey? extension button in the extension bar? I'll see.

No builds proper, but you should be able to extract the files and load unpacked extension. In Firefox:
![image](https://user-images.githubusercontent.com/12505802/172960754-7d77c5ad-5aee-45b1-8a86-b54b867a9533.png)
(You need to select manifest.json)

Chrome: this switch goes to 'yes' and then you open the folder with files:
![image](https://user-images.githubusercontent.com/12505802/172960512-378d0fad-15b3-470c-807a-90c057c33820.png)

The zips with the files:
* [dist-ff.zip](https://github.com/tamius-han/ultrawidify/files/8874456/dist-ff.zip)
* [dist-chrome.zip](https://github.com/tamius-han/ultrawidify/files/8874455/dist-chrome.zip)


----------

For those who stumble upon this and just want dem screenshots

# Current work in progress
This is how the new interface will roughly look like:
![image](https://user-images.githubusercontent.com/12505802/172956334-dc05994a-eb89-4663-bde7-9922b00aa4f2.png)

I still have to figure out what to do with firefox, because firefox doesn't do `background-filter` yet.

When the popup opens, there'll be website + status, which will let you know how much support you can expect:

## Support level badge

If you're on one of the sites that I pay attention to, there'll be verified badge:
![image](https://user-images.githubusercontent.com/12505802/172956736-6e66c768-1913-4e5d-9f13-3e0ad8062913.png)

There's also
* Community — (sites that people say extension works on (or that people help me get extension to work on), because subscription services be expensive and geoblocked)
* Unknown — best effort/you're on your own
* You're the boss — for sites that user configures manually

![image](https://user-images.githubusercontent.com/12505802/172956475-7a72fd81-5305-4e19-9671-ff9a9ca4eaf1.png)
Which is the reason I went with Twitch example.

## Adding aspect ratios and changing keyboard shortcuts

Right there, in the UI, there's now a very easy to miss button:
![image](https://user-images.githubusercontent.com/12505802/172957775-f91c1263-b3c3-4c78-af47-c0e3f60cb55f.png)

Click it and you get new and non-sucky interface for configuring shortcuts. Not only is this no longer is this hidden away on settings page, you also no longer have to sacrifice a few d20s to satan and chant your D&D rulebooks backwards in order for things to work. 
![image](https://user-images.githubusercontent.com/12505802/172957931-13412588-825c-4566-aad6-e729676d329e.png)

This includes adding new aspect ratios:
![image](https://user-images.githubusercontent.com/12505802/172958328-9f202fa1-d128-4652-baae-a95050d7cf67.png)

Also applicable to 'stretch video' options.

## The zoom slider is back:
![image](https://user-images.githubusercontent.com/12505802/172958438-d023a8fa-291d-4d67-aa0e-16435ac4fb69.png)

You can also unlock aspect ratio when zooming (though the second slider initializes a bit broken):
![image](https://user-images.githubusercontent.com/12505802/172958521-e20a924f-2de0-45c3-a03d-704be4830f1d.png)

## Alignment options:
![image](https://user-images.githubusercontent.com/12505802/172958580-40861868-47d1-4561-8be9-d0f0528bcf96.png)
Yes, I've heard that one feature request. No longer is alignment limited to left-center-right, you also get to align to top and bottom.

## Player detection:
Every now and then, there's gonna be someone coming in with "hey this site doesn't work (anymore)" for some site that I, for one reason or another, cannot access (geo blocking, subscriptions).  This usually happens because player element (aka the zone in which video is supposed to be playing) was not detected, or was not detected correctly.

Ultrawidify currently does have interface that allows users to override player element detection, but it requires you to actually know how websites work:
![Screenshot_20220610_010545](https://user-images.githubusercontent.com/12505802/172959621-6d5fe18a-5aca-4d28-b877-4826978cea19.png)

And that's not cool because an average person doesn't know what's a query selector and what to put in those boxes.

My usual way of dealing with that so far is asking people to install [this extension] and save the webpage with [said extension] and send the HTML file to me, with most of the resources (normally 'save page as' doesn't save those).

However, that approach only worked in ~50% of the time (some sites get massively broken with that extension, still) ... so I figured out I'll try to have something that would help people to pick the player for themselves. This is WIP AF at the moment:

![image](https://user-images.githubusercontent.com/12505802/172958707-2a9e54e3-a135-4e7e-a133-f69cd1d5b1dc.png)

The idea is that you get to hover down the list of elements with the mouse, until you find something that covers the area you expect your video to play in. And all that suffering happens because I want a one-size-fits-all approach that can deal with theater mode(s). 

This page currently does nothing yet, but with some extra functionality and lipstick I think it's gonna be worth it.

**Reply by Insan1ty0ne** [reply]
The reason that your ultrawide extrension is better than any other is that not only does it stretch video to 21:9 / 32:9, but it also CORRECTLY resizes the video to make sure that nothing gets cut off and that everything that was originally visible on-screen STAYS on screen. All of the other "ultrawide" extensions that I have used just zoom the video in the fill the screen at 21:9 and that causes all of the video around the edges to be cut off.

I am now using your latest commit to the feature/player-ui branch as an unpacked extension on the latest /Dev branch of Google Chrome and it works great. The only thing I can't figure out how to do is close the settings menu after applying the video settings I want. There isn't any exit button that I can see. So now I am just stuck with Ultrawidify open over top of any video I am watching!

**Reply by tamius-han** [reply]
Yea, kinda can't (except if you delete it away with devtools, or at least that was the situation last week).


_However_, lil bit of an update — I did manage to get the "planned stuff" from the previous comment working. Interface now gets triggered by hovering at the right spots and hidden by moving the mouse away from popup.

[dist-ff.zip](https://github.com/tamius-han/ultrawidify/files/8913576/dist-ff.zip)
[dist-chrome.zip](https://github.com/tamius-han/ultrawidify/files/8913575/dist-chrome.zip)

**Reply by Insan1ty0ne** [reply]
I can confirm that the setting interface now shows up properly and disappears properly when you are / aren't interacting with the video player. Definitely a big improvement from the last build because now I can daily drive this plugin again. The only thing I would say is that having it always pop-up every time you seek the player or move your mouse in fullscreen can get pretty annoying. Perhaps in the future you could have the interface pop-up function "sleep" after a user has set their preferred setting? I'm not sure how most users prefer to use the plugin but for me I only set it once when I first visit the page and then I don't need to set it again unless I change websites / etc. Keep up the great work!

**Reply by Insan1ty0ne** [reply]
> Yea, kinda can't (except if you delete it away with devtools, or at least that was the situation last week).
> 
> _However_, lil bit of an update — I did manage to get the "planned stuff" from the previous comment working. Interface now gets triggered by hovering at the right spots and hidden by moving the mouse away from popup.
> 
> [dist-ff.zip](https://github.com/tamius-han/ultrawidify/files/8913576/dist-ff.zip) [dist-chrome.zip](https://github.com/tamius-han/ultrawidify/files/8913575/dist-chrome.zip)

When you get time you should post updated distributions again for FF and Chrome with the most recent fixes. The Chrome App Store version still doesn't work for me but this branch works great.

**Reply by tamius-han** [reply]
Well beta branch going from beta to stores is going to take some time. 

Since I'm currently dealing with what should have been a performance improvement not going entirely according to plan on the master branch anyway ... 

 * Any specifics on what doesn't work on FF/Chrome branch?
 * Can I get the settings file (extension options page -> export settings)?
 * Does the issue get fixed if you reset the settings? (Reset settings after you've exported them)

Current list of show-stoppers:

**Feature parity department**

* Extension popup needs to be remade. The remake includes mostly deleting stuff (yay), but there's also a few things that need changes (courtesy of new event bus). I'll probably redo most of it from scratch, because it will lead to less spaghetti. 
  
  The features it needs (that need to be rewritten):
  * Site whitelisting/blacklisting (including frames)
  * Force-show UI on sites that, in the opinion of content script, don't contain videos (Disney+ I'm looking at you)
  
  Since those will be pretty much the only two things that the popup will do in the new major version, popup re-design is in order.
  
* Currently, in-player UI requires a `<video>` element, an element that extension determines is a div belonging to the player UI, and it requires that the player element is bigger than a certain size. When extension UI is force-shown, it needs to be sticked to the page element.
  * find a way to make this work with iframes
  * if no video is detected on the page, force-shown player UI should also default to player detect screen, with a big "detect video" button
* Make current "custom CSS" options compatible with settings that the new ever-so-slightly-user-friendlier player detect options spit out
* Move logging interface to in-player UI
* Pan with mouse while holding shift
 
**Major Goals of 6.0 department**

* At least basic autodetection settings (primarily frequency and strictness) in player UI.
* When you change settings for player detection and/or autodetection, actually _save_ them
* Manual video alignment: provide some draggable UI, not just "align to corners"
* Manual video alignment: option to account for current crop when aligning video. Make that the default. This one could be problematic a little.
* There's been some changes in how settings are formatted between 5.x and 6.x, so settings migration will have to be figured out

**Optional goals of 6.0**

* Configuration options for in-player UI (trigger area location, size; popup location, size)
* Extension settings page:
 * Have a sub-page with all whitelisted/blacklisted sites
 * ExtensionConf has changed. Extension settings page needs to either be updated with the changes, or — more likely — will vanish since most of the stuff should be configurable from the popup.

**Other quality of life:**
* manifest 3 (required by the end of the year)
* with manifest 3, Chrome now finally supports [tabs.removeCSS](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/tabs/removeCSS), except it's called [scripting.removeCSS](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/scripting/removeCSS) now. Probably not a biggie.
* When site is blacklisted, disable it _immediately_ without requiring page reload (this will be possible now that removeCSS is present on both browsers)

**Nice-to-haves**

* 'Revert to default settings' option on sites where user has meddled with the settings
* Visual debugging tools for autodetection, to see what parts of an image are tripping autodetection.
* Better, more accessible and more up-to-date documentation ~~that nobody will read anyway~~
* Rewrite of README.MD to account for redesign
* boy if i could find some time to get jenkins back up and running

And this is probably not a complete list.
